### PR TITLE
Optimize ReaderWriter

### PR DESCRIPTION
**Pull request by IveanEx**
This PR optimizes the following in ReaderWriter: 

<s>Replace the **spatialAddrGen** with full adder implementation</s>
Replace all buffers with pipe buffers (except the scalatest one)

### CODE REVIEW

**Comment by IveanEx** [comment]
@rgantonio If you are free, please check if the critical path in control path (valid / ready hand shaking signal) deteriorates with this improvements 😊😊

**Review by xiaoling-yi — APPROVED** [review]

File: hw/chisel/src/test/scala/snax/readerWriter/ReaderTester.scala.bak
Comment (rgantonio) [inline hw/chisel/src/test/scala/snax/readerWriter/ReaderTester.scala.bak:1]: Hmm please don't put backups in the repo. Leave them be in the local computers.

**Review by rgantonio — CHANGES_REQUESTED** [review]
Setting barrier here first before merge hehe. So that we make sure that the necessary changes are indeed helpful 😄 

Wait a few hours hold on. I will test.

**Comment by rgantonio** [comment]
Also, did you trace or check in Chisel if it fixes the long path?

**Review by rgantonio — APPROVED** [review]
Ok! I can test tomorrow but was this pipe thing from the queue the fix? I can test in synthesis first before plugging in.

**Pull request merged**